### ISSUE TITLE
Functions lost when serializing object, no warnings on deserialization.

### ISSUE THREAD

**Issue by homer-jay** [palantir/tslint#3961]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.9.1
- __Running TSLint via__:VSCode

#### TypeScript code being linted

```ts
class Greeter {
    greeting: string;
    constructor(message: string) {
        this.greeting = message;
    }
    greet() {
        return "Hello, " + this.greeting;
    }
}

let greeter = new Greeter("world");
let anotherGreeter: Greeter = JSON.parse(JSON.stringify(greeter));
try {
    console.debug(anotherGreeter.greet());
} catch (e) {
    alert(e);
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "no-duplicate-variable": true,
    "no-unused-variable": [
      true
    ]
  },
  "rulesDirectory": [
    "node_modules/tslint-eslint-rules/dist/rules"
  ]
}
```

#### Actual behavior
The code throws an error: TypeError: anotherGreeter.greet is not a function. No warnings are produced by TSLint. JSON.serialize has only conserved the properties, and has removed the functions.

#### Expected behavior
You would expect a warning whenever JSON.parse, which returns any, is assigned to a variable of another more specific type. The code I've posted is too obvious, but on practice it is very common to serialize on one place and deserialize on another place, and it shows that it is possible to write unsafe code without getting any warning.

**Comment by giladgray** [comment]
if i'm not mistaken, you cannot serialize functions: what is the **string representation** of a function??

this is simply a language limitation.

**Comment by giladgray** [comment]
also, `JSON.parse` returns `any`: no knowledge of the original data type (the way it should be).

**Comment by homer-jay** [comment]
@giladgray The string representation of a function is the text of that snippet. It could be serialized and TS could provide an overriden version of JSON.parse and JSON.serialize, but that is more a language thing and is not the purpose of this issue.
As for the linter, OFC JSON.parse return any, thats why it should give a warning when assigning it to something else !== any. Unsafe cast warnings are common in other languages.
TL;DR: Consider adding an unsafe cast warning.
